---
title: "Check HS Generation GWAS"
author: "Daniel M. Gatti, Ph.D."
date: "8/26/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

library(knitr)
library(tidyverse)
library(qtl2)
library(qtl2convert)

# Generation that we used for each haplotype reconstruction.
gen_names = c(2, 25, 50)

base_dir = '/media/dmgatti/data1/ColoState/ATM'
data_dir = file.path(base_dir, 'data')
fig_dir  = file.path(base_dir, 'figures')
geno_dir = file.path(data_dir, 'genotypes')
hap_dir  = file.path(base_dir, 'haplo_reconstr')
results_dir  = file.path(base_dir, 'results')
probs_files = file.path(hap_dir, str_c('atmko_hs_alleleprobs_cleaned_20pct_129_gen', gen_names,'.rds'))
pheno_file  = file.path(data_dir, 'phenotypes', 'amtko_hs_phenotypes_cleaned.csv')
  
muga_dir     = '/media/dmgatti/data0/MUGA/'
gm_wisc_file = file.path(muga_dir, 'gm_uwisc_v1.csv')
hs_snp_file  = file.path(muga_dir, 'atmkohsnpt_variants_129S1_SvImJ.sqlite')

founder_names = c('A/J', 'AKR/J', 'BALB/cJ', 'C3H/HeJ', 'C57BL/6J', 'CBA/J', 'DBA/2J', 'LP/J', '129S6/SvEvTac-Atm/J')

hs_founder_prop_file = '/media/dmgatti/data1/ColoState/HS/results/genome_founder_prop.csv'

HScolors = CCcolors
HScolors = c(CCcolors, '129SvE' = '#BB5500')
HScolors[1] = '#FFC800'
names(HScolors) = founder_names

snp_func = qtl2::create_variant_query_func(hs_snp_file)

rankZ = function(x) {
  x = rank(x, na.last = "keep", ties.method = "average") / (sum(!is.na(x)) + 1)
  return(qnorm(x))
}
```

Map survival using the genoprobs from the four different 129S proportions: 20, 30, 40 & 50%.

Read in each probs file and write out the GWAS results.

Load in genoprobs, map and pheno.

```{r read_data}
pheno = read.csv(pheno_file)
rownames(pheno) = pheno$mouse
pheno$log_days = log(pheno$days)
pheno$rz_days  = rankZ(pheno$days)

markers = read.csv(file = gm_wisc_file)
markers = markers[,c('marker', 'chr', 'bp_mm10')]
colnames(markers)[3] = 'pos'
markers[,3] = markers[,3] * 1e-6
markers = subset(markers, chr %in% c(1:19, 'X'))
map = map_df_to_list(markers, pos_column = 'pos')

```

Synch up samples and markers.

```{r synch_samples_markers}
for(i in 1:length(probs_files)) {
  
  gen = gen_names[i]
  print(paste("Generation:", gen))

  probs = readRDS(probs_files[i])

  samples = intersect(pheno$mouse, rownames(probs[[1]]))
  setdiff(pheno$mouse, rownames(probs[[1]]))
  setdiff(rownames(probs[[1]]), pheno$mouse)

  pheno = pheno[samples,]
  probs = lapply(probs, function(z) { z[samples,,] })
  for(j in seq_along(probs)) {
    map[[j]] = map[[j]][dimnames(probs[[j]])[[3]]]
  } # for(j)

  sapply(map, length)
  sapply(probs, dim)[3,]

  # I'm not going to set any covariates right now since I see no sex difference
  # in survival.

  # Set attributes on probs that didn't get set.
  attr(probs, 'crosstype') = 'genail9'
  attr(probs, 'is_x_chr') = setNames(c(rep(F, 19), T), c(1:19, 'X'))
  attr(probs, 'alleles') = LETTERS[1:9]
  attr(probs, 'alleleprobs') = TRUE
  attr(probs, 'class') = c('calc_genoprob', 'list')

  K = qtl2::calc_kinship(probs, type = 'loco', cores = 4)

  allK = qtl2::calc_kinship(probs, type = 'overall', cores = 4)
  print("Heritability")
  print(est_herit(pheno = pheno[,'log_days',drop = F], kinship = allK))
  
  gwas = scan1snps(genoprobs = probs, pheno = pheno[,'log_days',drop = F], 
                   map = map, kinship = K, keep_all_snps = FALSE, 
                   query_func = snp_func, cores = 8)
  
  png(file.path(fig_dir, paste0('log_survival_', gen, 'gen.png')), width = 1000, height = 800, res = 128)
  plot_snpasso(gwas$lod, gwas$snpinfo, main = paste('Generation', gen, 'Log Survival'))
  dev.off()
  
  print(paste('Max LOD (log):', max(gwas$lod)))
  
  gwas = scan1snps(genoprobs = probs, pheno = pheno[,'rz_days',drop = F], 
                   map = map, kinship = K, keep_all_snps = FALSE, 
                   query_func = snp_func, cores = 8)
  
  png(file.path(fig_dir, paste0('rz_survival_', gen, 'gen.png')), width = 1000, height = 800, res = 128)
  plot_snpasso(gwas$lod, gwas$snpinfo, main = paste('Generation', gen, 'RankZ Survival'))
  dev.off()
  
  print(paste('Max LOD (rz):', max(gwas$lod)))

} # for(i)
```

# Count number of recombinations and look at the number of 129S blocks.

```{r count_recomb}
# Get the 36 state genoprobs files.
probs_files = file.path(hap_dir, str_c('atmko_hs_genoprobs_20pct_129_gen',  gen_names, '.rds'))

count_crossovers = function(pr) {
  
  maxgt = apply(pr, c(1, 3), which.max)
  maxgt = matrix(colnames(pr)[maxgt], nrow = nrow(maxgt), ncol = ncol(maxgt),
                 dimnames = list(rownames(maxgt), colnames(maxgt)))
  return(rowSums(maxgt[,-1] != maxgt[,-ncol(maxgt)]))

} # count_crossovers()

count_129_prop = function(pr) {
  
  pr = pr[,grep('I', colnames(pr)),]
  pr = apply(0.5 * pr[,1:8,], c(1, 3), sum) + pr[,9,]
  return(rowMeans(pr))
  
} # count_129_prop()

xo_list   = vector('list', length(probs_files))
p129_list = vector('list', length(probs_files))
names(xo_list)   = gen_names
names(p129_list) = gen_names

for(i in 1:length(probs_files)) {
  
  gen = gen_names[i]
  print(paste("Generation:", gen))

  probs = readRDS(probs_files[i])

  xo_list[[i]]   = lapply(probs, count_crossovers)
  p129_list[[i]] = lapply(probs, count_129_prop)
    
} # for(i)
```

Gather results by sample.

```{r}
n_samples = length(xo_list[[1]][[1]])
n_chr = length(xo_list[[1]])

xo_smry = vector('list', length(xo_list))
names(xo_smry) = names(xo_list)
p129_smry = vector('list', length(p129_list))
names(p129_smry) = names(p129_list)

for(i in seq_along(xo_list)) {
  xo_smry[[i]] = matrix(unlist(xo_list[[i]]), nrow = n_samples, ncol = n_chr, 
                        dimnames = list(names(xo_list[[1]][[1]]), names(xo_list[[1]])))
  xo_smry[[i]]   = rowSums(xo_smry[[i]])
  
  # p129_smry[[i]] = lapply(p129_list[[i]], rowMeans)
  p129_smry[[i]] = matrix(unlist(p129_list[[i]]), nrow = n_samples, ncol = n_chr,
                          dimnames = list(names(xo_list[[1]][[1]]), names(xo_list[[1]])))
  p129_smry[[i]] = rowMeans(p129_smry[[i]])
  
} # for(i)
```

Plot results.

```{r plot_num_crossovers}
boxplot(data.frame(xo_smry), main = 'Number of Crossovers per sample')
```

```{r plot_prop_129}
boxplot(data.frame(p129_smry), main = 'Proportion of 129 genome')
```

```{r}
sapply(xo_smry, median)
```

```{r}
sapply(p129_smry, median)
```
It looks like 20% is the value that gives the best set of metrics: crossovers near 270 and % 129 near 54%.
